java/arcs/core/util/Lens.kt

```diff
@@ -0,0 +1,131 @@
+package arcs.core.util
+
+/**
+ * A miniature lightweight "Optics" utility for Arcs. Mostly used to modifying immutable graph
+ * data structures used by the [Allocator].
+ *
+ * A lens is an kind of optic in functional programming that allows immutable data structures to
+ * be updated compactly and elegantly. A [Lens] is constructed around a data structure and
+ * allows its focus target (e.g. field member of a type) to be fetched, modified, or set.
+ *
+ * @property getter a lambda which given [Target] returns [Focus]
+ * @property setter a lambda which given [Target] and [Focus] returns new [Target] set with [Focus]
+ *
+ * Based on the ideas in
+ * http://davids-code.blogspot.com/2014/02/immutable-domain-and-lenses-in-java-8.html
+ */
+open class Lens<Target, Focus>(
+    private val getter: (Target) -> Focus,
+    private val setter: (Target, Focus) -> Target
+) {
+    /** Invokes a getter on the [Target] and returns a [Focus] value. */
+    fun get(target: Target) : Focus = getter(target)
+
+    /** Invokes a setter on the [Target] with the given [Focus] value and returns a new [Target]. */
+    fun set(target: Target, value: Focus) : Target = setter(target, value)
+
+    /**
+     * Applies a function to the value of invoking [get] on the [Target], and then invokes
+     * [set] on the [Target] with the value, returning a new [Target].
+     */
+    fun mod(target: Target, f: (Focus) -> Focus): Target {
+        return set(target, f(get(target)))
+    }
+
+    /**
+     * Allows two or more [Lens] to be composed and chained so that a deeply nested immutable
+     * data structure can be elegantly updated. Returns a new [Lens].
+     */
+    fun <Parent> comp(parentLens: Lens<Parent, Target>) : Lens<Parent, Focus> =
+        object : Lens<Parent, Focus>(
+            { parent : Parent -> get(parentLens.get(parent)) },
+            { parent : Parent, focus: Focus ->
+                parentLens.mod(parent) { target: Target ->
+                    set(target, focus)
+                }
+            }
+        ) {}
```

Comment:
are the brackets needed?

Reply:
Apparently, it gives the error 'expecting a class body' :(